src/coredump/coredump.c

```diff
@@ -1342,14 +1346,68 @@ static int gather_pid_metadata_from_argv(
                         t = buf;
                 }
 
+                if (i == META_ARGV_PID) {
+                        /* Store this so that we can check whether the core will be forwarded to a container
+                         * even when the kernel doesn't provide a PIDFD. Can be dropped once baseline is
+                         * >= v6.16. */
+                        r = pidref_set_pidstr(&kernel_pidref, t);
+                        if (r < 0)
+                                return log_error_errno(r, "Failed to initialize pidref from pid %s: %m", t);
+                }
+
+                if (i == META_ARGV_PIDFD) {
+                        assert(!pidref_is_set(&context->pidref));
+
+                        if (!isempty(t)) {
+                                assert(kernel_fd < 0);
+
+                                kernel_fd = parse_fd(t);
+                                if (kernel_fd < 0)
+                                        return log_error_errno(kernel_fd,
+                                                               "Failed to parse pidfd \"%s\": %m",
+                                                               t);
+
+                                r = pidref_set_pidfd(&context->pidref, kernel_fd);
+                                if (r < 0)
+                                        return log_error_errno(r,
+                                                               "Failed to initialize pidref from pidfd %d: %m",
+                                                               kernel_fd);
+                        } else {
+                                /* If the kernel didn't give us a PIDFD, then use the one derived from the
+                                 * PID immediately, given we have it, and skip the metadata field. */
+                                context->pidref = TAKE_PIDREF(kernel_pidref);
```

Comment:
move this out of the loop. i.e. once the loop is done do:

```c
if (!pidref_is_set(&context->pidref))
        context->pidref = TAKE_PIDREF(kernel_pidref);
```
because otherwise, if the pattern is called without the pidfd parameter at the end, then we'll still not use a pidfd at all, but I think we should. i.e. our locally allocated pidfd is still better than none

Reply:
but the pidfd that pidref_set_pidstr() allocates here is closed, and then another one allocated. I want that we just keep the first one we allocate around, until we are completely done.

if we open and close and open and close and open and close pidfds to the same process all the time, we might just not do it at all